PM - Get rid of some weak functions

ceolin opened the pull request:
This pr changes how pm policies and notifications are done. Instead of using weak functions that can be override by the application this pr defines new types and APIs for that. I kept the same behavior and name conventions for now, the idea is to do incremental steps. Currently I took slightly different approaches between the notification and policy API. For the policy, the API allows application to extend the basic policy, that is a little bit different from what we have being done so, any feedback is welcome here.

The next step, is to define a better name convention for power management, organize headers, power states, ... (ensure that points agreed in https://github.com/zephyrproject-rtos/zephyr/pull/27315 are addressed)

tbursztyka commented on doc/reference/power_management/index.rst:
```diff
@@ -153,12 +153,41 @@ Kconfig option) less than or equal to the scheduled system idle time duration.
+         /* Custom fields */
+   };
+
+Finally, a policiy type must be defined using
```
typo

pabigot replied:
also next line

tbursztyka commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+/**
+ * @brief Function to create device PM list
+ */
+void sys_pm_create_device_list(void);
```
hopefully this will disappear someday (list will be build time generated etc...), and it won't be necessary to mention any "PM device list" like below.

pabigot replied:
The list of devices participating in power management cannot be static: devices may be enabled or disabled, or even (in the future) instantiated at runtime.

tbursztyka replied:
disabled (or non initialized) would be silently skipped from the whole dev list. Seems simpler that way than creating a list at runtime each time you need to do something around pm.

pabigot replied:
Agreed: having PM capture and maintain a list of devices doesn't make sense, unless the cost of skipping over devices that don't support PM becomes so excessive that the cost of maintaining a separate list is worth doing.  I'd start assuming it's not worth it.

ceolin replied:
agreed, this is not part of the policy though. I didn't touch it yet, just move it around.

tbursztyka commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	if (sys_pm_policy->next_state_get == NULL) {
+		/* Should it be -ENOSYS ? */
```
-ENOTSUP seems fine to me

tbursztyka replied:
ok

ceolin replied:
Initially I had used ENOSYS, but then in checkpatch.pl we have "ENOSYS means 'invalid syscall nr' and nothing else\n", that is not the case here and I saw some places using ENOTSUP. I think ENOSYS is better, but I tried to be more consistent with the rest of the code base.

tbursztyka replied:
The issue with checkpatch is that it's entirely Linux centric, which generates troubles as this one sometimes.

tbursztyka commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	if (sys_pm_policy->low_power_devices_get == NULL) {
+		/* Should it be -ENOSYS ? */
```
same

pabigot commented on include/power/power.h:
```diff
@@ -64,6 +64,43 @@ extern unsigned char sys_pm_idle_exit_notify;
+ * @param exit_fn  Function to be called when platform exits a power state
+ */
+#define PM_NOTIFIER_DEFINE(name, entry_fn, exit_fn)		\
+	static Z_STRUCT_SECTION_ITERABLE(pm_notifier, name) =	\
```
We should get some clarity on when objects are aggregated into an array by the linker, and when they are not.

If `pm_notifier` were a kernel object the aggregation would make sense.  Otherwise this API doesn't seem ideal, as it assumes that all notifiers are identified at build time and are enabled unconditionally at boot.  This forces dependencies: a module may need to set things up before its notifier is ready to be called, which won't have been done if the system does any suspension before the module is initialized (which may not happen at boot).

I think a register/unregister API that adds/removes a notifier is more flexible and has only a small cost (a link node in the notifier structure).

ceolin replied:
They can be initialized with NULL pointer callbacks in this case, we can go with the linked list approach, I have nothing against that. What I want here is confirm whether or not makes sense allow multiple notifiers.
Btw, make it kernel objects may make sense if applications want to use it from user mode and I have the feeling that this will be a valid case.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+void sys_pm_create_device_list(void);
+
+/**
+ * @brief Function to suspend the devices in PM device list
```
All these need to clearly define what "suspend", "low_power", "force_suspend" mean.

ceolin replied:
Intend to work on it when changing the power states. This API much likely will need to change, this pr is just adding the foundation for that.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+	extern struct pm_policy *sys_pm_policy;
+
+	PM_POLICY_FOREACH(pm_policy, policy) {
+		if (policy->name != NULL &&
```
is there a valid use for a policy that doesn't have a name, if they're identified by name?

ceolin replied:
Nope, it is just a sanitycheck

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @return false if the policy is not available, true otherwise.
+ */
+static inline bool sys_pm_policy_set(const char *policy_name)
```
What will be invoking this function?  Is it reasonable to do it by name, rather than passing a reference to the policy structure the application wants to use?

ceolin replied:
This is just more convenient for the application, instead of having to do:
```
extern struct pm_policy *dummy_policy;
sys_pm_policy_set(dummy_policy);
```

pabigot replied:
Why would the policy be extern?  Isn't the app providing the policy, so it'd not be an external global symbol, but a static local one?  If there are default policies available to reference they should be declared in a header somewhere.

ceolin replied:
The idea is that the symbol is implicitly created by the macro, and can be in a different object file, not necessarily in the same  file that the application will configure the policy, this is the same behavior of other similar macros.

Though as we had discussed in a previous meeting, there is no support for multiple policies  anymore, so this is not relevant anymore.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+ */
+static inline bool sys_pm_policy_set(const char *policy_name)
+{
+	extern struct pm_policy *sys_pm_policy;
```
why does this have to be global?  It should probably be `z_sys_pm_policy` if it is, to indicate nobody except system PM is to touch it.

ceolin replied:
yep, that makes sense.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+	struct { \
+		const char *name; \
+		const size_t len; \
+		enum power_states (*next_state_get)(struct pm_policy *policy,\
```
I think the first argument should be `const struct pm_policy *`.  There should be no reason to have mutable data in the policy structure: if it needs state, it should store it somewhere else.

ceolin replied:
I was in doubt about just adding a pointer that can be used by the implementation to store its context or allow extending the type. I can change it, in the end it will not increase the struct size because I can get rid of `len` and it will simplify some things.

pabigot reviewed (changes requested):
We should get some clarity on when objects are aggregated into an array by the linker, and when they are not.

If `pm_notifier` were a kernel object the aggregation would make sense.  Otherwise this API doesn't seem ideal, as it assumes that all notifiers are identified at build time and are enabled unconditionally at boot.  This forces dependencies: a module may need to set things up before its notifier is ready to be called, which won't have been done if the system does any suspension before the module is initialized (which may not happen at boot).

I think a register/unregister API that adds/removes a notifier is more flexible and has only a small cost (a link node in the notifier structure).

This may or may not also apply to policies, but I'd like to see a more clear description of what it would mean to dynamically add/remove/change the system power policy.  The system state under one policy would be unlikely to support the needs of another policy, and without seeing the events that produced it the new policy may not be able to accurately reconstruct the system state.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,183 @@
+ * Store information about a power management policy.
+ */
+struct pm_policy {
+	PM_POLICY_BASE;
```
While we're talking: I'd really rather not see the common parts of the structure provided by a define.  It should be a standalone structure, which will be the first member of  a specific policy structure, or accessed by `CONTAINER_OF`.  Field declarations in a macro are really hard to read.

ceolin replied:
I just changed to something simpler ... Just use a basic type wouldn't work as I wanted, since it was supposed to be statically allocated by the `PM_POLICY_DEFINE` macro. I simplified it and now it is not possible to extends the type.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,118 @@
+ * Store information about a power management policy.
+ */
+struct pm_policy {
+	enum power_states (*next_state_get)(struct pm_policy *policy,
```
The signature has changed to add an error, but as a parameter.  For consistency with other API this should return an `int` success/error code, and store the next state as an output parameter.

pabigot commented on doc/reference/power_management/index.rst:
```diff
@@ -153,12 +153,23 @@ Kconfig option) less than or equal to the scheduled system idle time duration.
-the following function.
+The power management policy is define by the application implementing
+the methods declared in :code:`struct pm_policy` and declared using
+:code:`PM_POLICIY_DEFINE()` macro.
```
s/POLICIY/POLICY/

pabigot commented on include/power/power.h:
```diff
@@ -64,6 +64,26 @@ extern unsigned char sys_pm_idle_exit_notify;
+ * Power managment notifier struct
+ *
+ * This struct contains callbacks that are called when the target enters and
+ * exits power states.
```
We need documentation here, and/or at the register/unregister functions, that explains when the set of notifiers can be changed.  The code iterates over the list without any protection for the list changing as a result of these functions being invoked.

ceolin replied:
thinking about it, that is trick, the list is accessed from ISR as well, what makes mutex nonviable. If I use a semaphore I can't wait if its unavailable. In this case if the semaphore is not available I can't access the list and consequently not send notifications. The other problem is that suspend is not only called from the idle thread, it can be directly called from the application. Ideas ?

pabigot replied:
I don't see how this has been resolved, so unresolving it.

AFAICT these are invoked from `_sys_suspend` and `_sys_resume`.  Are those from interrupt context?  Do the notification callbacks have to be invoked from that context, or can/should they be pushed to a work queue?

ceolin replied:
@pabigot Forgot to document it. I just protected the list of notifiers with a k_spin_lock to avoid race condition when accessing these structs. Will document that these callbacks can be called from an interruption context, at this point I don't know if that is problem, so I'll keep this behavior since that is how is currently done.

In the future if needed, I change it to a workqueue. This won't change the API anyway.

pabigot reviewed (commented):
A few additional comments based on current status.  Some of my previous ones may still be relevant.

ceolin commented:
> A few additional comments based on current status. Some of my previous ones may still be relevant.

Thanks for reviewing it. I'll check them again, but I believe they were either addressed or became invalid with latest changes

pabigot reviewed (changes requested):
Needs to be rebased now that things have been renamed.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,114 @@
+/**
+ * @brief Function to put the devices in PM device list in low power state
+ */
+int pm_low_power_devices(void);
```
It would be much easier to review this if the renaming of the original version of this file were done in a separate commit from the changes to the file content.  It's not obvious what's new, and what's just refactoring existing API.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,114 @@
+ */
+#define PM_POLICY_DEFINE(policy_name, policy_next_state_get,		\
+			 policy_low_power_devices_get)			\
+	const struct pm_policy policy_name = {				\
```
This should probably be `static` unless there's a reason to make it globally visible.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,114 @@
+
+	if ((z_sys_pm_policy) == NULL ||
+	    (z_sys_pm_policy->next_state_get == NULL)) {
+		return -ENOSYS;
```
Should be `-ENOTSUP`.  See #23727.

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,114 @@
+/**
+ * @brief Function to get the next PM state based on the ticks
+ *
+ * @return 0 in case of success, a negative value otherwise
```
Please document the known values.  It's fine to have a fallback for future-proofing.  E.g.

```
 * @return 
 *   * 0 in case of success
 *   * -ENOTSUP if there is no policy, or the police dpes not support a next-state capability;
 *   * other negative error codes indicating specific failure modes
```

pabigot commented on include/power/pm_policy.h:
```diff
@@ -0,0 +1,114 @@
+
+	if ((z_sys_pm_policy == NULL) ||
+	    (z_sys_pm_policy->low_power_devices_get == NULL)) {
+		return -ENOSYS;
```
`-ENOTSUP`.

pabigot reviewed (changes requested):
I'd appreciate a split of the move from the content change.  I don't see much significant here, so if it's mostly moving things around the review could be shorter.  There does appear to be some new capability that could be done another way.

Commit:
power: Move APIs away from policy

Move functions that don't belong to the policy manager to the power
subsystem.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Move pm_policy header to include dir

pm_policy is a public header that define APIs that can be implemented
by applications and SoCs. Just move it to include/power

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Get rid of weak functions used in policy

Currently the pm_policy is done overriding weak symbols. There are
some problems with this approach:

1) Undefined behavior in case of having two definitions of the same
   weak symbol.
2) Without an associated object it is not possible to control access
   through user mode.

This pr aims to address these problems, note that it is not adding a
new kernel object but it creates the foundation to do it if necessary.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
